subprojects/packagefiles/gmp/meson.build

```diff
@@ -0,0 +1,548 @@
+  mpn_extra_functions += ['invert_limb_table']
+endif
+
+limb = []
+if host_machine.system() == 'cygwin'
+  calling_conventions_objs = []
+  limb = ['longlong']
+  conf.set('HOST_DOS64', 1)
+endif
+
+if assembly and calling_conventions_objs.length() > 0
+  conf.set('HAVE_CALLING_CONVENTIONS', 1)
+endif
+
+cc = meson.get_compiler('c')
+
+# If we're supposed to be using a "long long" for a limb, check that it works.
+if limb.get(0, '') == 'longlong'
+  if not cc.compiles(
+    '''
+    long long  foo;
+    long long  bar () { return foo; }
+    int main () { return 0; }
+    '''
+  )
+    error('the chosen compiler does not support long long')
+  conf.set('DEFN_LONG_LONG_LIMB', '#define _LONG_LONG_LIMB 1')
+else:
+  conf.set('DEFN_LONG_LONG_LIMB', '/* #undef _LONG_LONG_LIMB */')
+endif
+
+fat_path = []
+fat_functions = []
+fat_thresholds = []
+if fat
+  mpn_search_path = ['x86_64/fat', 'x86_64']
+  fat_path = [
+    'x86_64',
+    'x86_64/fat',
+    'x86_64/k8',
+    'x86_64/k10',
+    'x86_64/bd1',
+    'x86_64/bt1',
+    'x86_64/bt2',
+    'x86_64/zen',
+    'x86_64/pentium4',
+    'x86_64/core2',
+    'x86_64/coreinhm',
+    'x86_64/coreisbr',
+    'x86_64/coreihwl',
+    'x86_64/coreibwl',
+    'x86_64/skylake',
+    'x86_64/atom',
+    'x86_64/silvermont',
+    'x86_64/goldmont',
+    'x86_64/nano',
+  ]
+  fat_functions = [
+    'add_n',
+    'addmul_1',
+    'bdiv_dbm1c',
+    'com',
+    'cnd_add_n',
+    'cnd_sub_n',
+    'copyd',
+    'copyi',
+    'dive_1',
+    'divrem_1',
+    'gcd_11',
+    'lshift',
+    'lshiftc',
+    'mod_1',
+    'mod_1_1',
+    'mod_1_1_cps',
+    'mod_1_2',
+    'mod_1_2_cps',
+    'mod_1_4',
+    'mod_1_4_cps',
+    'mod_34lsub1',
+    'mode1o',
+    'mul_1',
+    'mul_basecase',
+    'mullo_basecase',
+    'pre_divrem_1',
+    'pre_mod_1',
+    'redc_1',
+    'redc_2',
+    'rshift',
+    'sqr_basecase',
+    'sub_n',
+    'submul_1',
+  ] + [  # 64-bits-specific functions
+    'addmul_2',
+    'addlsh1_n',
+    'addlsh2_n',
+    'sublsh1_n'
+  ]
+  fat_thresholds = [
+    'MUL_TOOM22_THRESHOLD',
+    'MUL_TOOM33_THRESHOLD',
+    'SQR_TOOM2_THRESHOLD',
+    'SQR_TOOM3_THRESHOLD',
+    'BMOD_1_TO_MOD_1_THRESHOLD',
+  ]
+endif
+
+new_path = []
+if get_option('nails')
+  foreach p : mpn_search_path
+    new_path += [p, p+'/nails']
+  endforeach
+  mpn_search_path = new_path
+endif
+
+if not assembly
+  mpn_search_path = ['generic']
+  conf.set('NO_ASM', 1)
+endif
+
+if get_option('default_library') != 'static'
+    add_project_link_arguments(
+      '-no-undefined',
+      '-Wl,--export-all-symbols'
+      language: ['c', 'cpp']
+    )
+endif
+
+check_declarations = [
+  ['fgetc', 'stdio.h'],
+  ['fscanf', 'stdio.h'],
+  ['optarg', 'unistd.h'],
+  ['ungetc', 'stdio.h'],
+  ['vfprintf', 'stdio.h'],
+  ['sys_errlist', 'errno.h'],
+  ['sys_nerr', 'errno.h'],
+]
+
+foreach f : check_functions
+  if cc.has_header_symbol(f.get(1), f.get(0))
+    conf.set('HAVE_DECL_' + f.get(0).to_upper(), 1)
+  endif
+endforeach
```

Comment:
really unsure about this bit

Reply:
In configure.ac it's this :

```m4
# Reasons for testing:
#   optarg - not declared in mingw
#   fgetc, fscanf, ungetc, vfprintf - not declared in SunOS 4
#   sys_errlist, sys_nerr - not declared in SunOS 4
#
# optarg should be in unistd.h and the rest in stdio.h, both of which are
# in the autoconf default includes.
#
# sys_errlist and sys_nerr are supposed to be in <errno.h> on SunOS according
# to the man page (but aren't), in glibc they're in stdio.h.
#
AC_CHECK_DECLS([fgetc, fscanf, optarg, ungetc, vfprintf])
AC_CHECK_DECLS([sys_errlist, sys_nerr], , ,
[#include <stdio.h>
#include <errno.h>])
```

Reply:
woops and there's even a typo

Reply:
Can you clarify what about it you aren't sure about?

Reply:
I'm looking at what `AC_CHECK_DECLS` does according to [this webpage](https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Generic-Declarations.html) and i'm not sure it's really equivalent of `cc.has_header_symbol()`